File: core/utils/src/commonMain/kotlin/de/stefan/lang/coreutils/contract/logging/Loggable.kt
package de.stefan.lang.coreutils.contract.logging

public interface Loggable {
    public val logger: Logging
    public val tag: String get() = this::class.simpleName ?: "Unknown"

    public fun logD(message: String) { logger.d(tag, message) }
    public fun logI(message: String) { logger.i(tag, message) }
    public fun logW(message: String) { logger.w(tag, message) }
    public fun logE(message: String) { logger.e(tag, message) }
}

File: iosApp/iosApp/core/logging/Loggable.swift
import Foundation
import shared

protocol Loggable {
    var logger: any Logger { get }
    var tag: String { get }

    func logD(message: String)
    func logI(message: String)
    func logW(message: String)
    func logE(message: String)
}

extension Loggable {
    var logger: any Logger { SharedModule.shared.logger()}
    var tag: String { TypeDescriptionUtil.typeName(from: self) }

    func logD(message: String) {
        logger.d(tag: tag, message: message)
    }

    func logI(message: String) {
        logger.i(tag: tag, message: message)
    }

    func logW(message: String) {
        logger.w(tag: tag, message: message)
    }

    func logE(message: String) {
        logger.e(tag: tag, message: message)
    }
}
